src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -659,6 +646,16 @@ public static void Run(ServiceBase[] services)
             }
             finally
             {
+                // Free the pointer to the name of the service on the unmanaged heap.
+                // We don't need to free the last element in the unmanaged array since the last element have null values (denotes the end of the table).
+                // Entries other than the last entry having a null pointer means that we failed while processing the entries for some reason
+                // and only elements up to this point has been successfully allocated so far (thus have to be deallocated)
+                for (int i = 0; i < entries.Length - 1 && entries[i].name != IntPtr.Zero; i++)
+                {
+                    Marshal.FreeHGlobal(entries[i].name);
+                }
```

Comment:
Is it correct that I do not have to do any clean-ups in regards with the function pointer obtained from `Marshal.GetFunctionPointerFromDelegate`? From what I've seen, the documents did not mention anything about clean-ups and it suggested that the function pointers are collected along with the delegate itself during GC, but just wanted to confirm.

Reply:
Correct. The functions pointer returned from `Marshal.GetFunctionPointerFromDelegate` is alive for as long as the delegate that it was created from is alive. There is no way to control the lifetime explicitly. (BTW: The implicit lifetime is very common source of bugs. It is not a very good design.)